File: qt/gen_qeventloop.go
package qt

import (
	"unsafe"
)

type QEventLoop__ProcessEventsFlag int

const (
	QEventLoop__AllEvents              QEventLoop__ProcessEventsFlag = 0
	QEventLoop__ExcludeUserInputEvents QEventLoop__ProcessEventsFlag = 1
	QEventLoop__ExcludeSocketNotifiers QEventLoop__ProcessEventsFlag = 2
	QEventLoop__WaitForMoreEvents      QEventLoop__ProcessEventsFlag = 4
	QEventLoop__X11ExcludeTimers       QEventLoop__ProcessEventsFlag = 8
	QEventLoop__EventLoopExec          QEventLoop__ProcessEventsFlag = 32
	QEventLoop__DialogExec             QEventLoop__ProcessEventsFlag = 64
	QEventLoop__ApplicationExec        QEventLoop__ProcessEventsFlag = 128
)

type QEventLoop struct {
	h          uintptr
	isSubclass bool
}

// NewQEventLoop constructs a new QEventLoop object.
func NewQEventLoop() *QEventLoop {
	g := newQEventLoop(QEventLoop_new())
	g.isSubclass = true
	return g
}

// NewQEventLoop2 constructs a new QEventLoop object.
func NewQEventLoop2(parent *QObject) *QEventLoop {
	g := newQEventLoop(QEventLoop_new2(parent.cPointer()))
	g.isSubclass = true
	return g
}

func (this *QEventLoop) MetaObject() *QMetaObject {
	return newQMetaObject(QEventLoop_MetaObject(this.h))
}

func (this *QEventLoop) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QEventLoop_Metacast(this.h, param1_Cstring))
}

func QEventLoop_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QEventLoop_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QEventLoop) ProcessEvents() bool {
	return (bool)(QEventLoop_ProcessEvents(this.h))
}

func (this *QEventLoop) ProcessEvents2(flags ProcessEventsFlags, maximumTime int) {
	QEventLoop_ProcessEvents2(this.h, flags, (int)(maximumTime))
}

func (this *QEventLoop) ProcessEvents3(flags ProcessEventsFlags, deadline QDeadlineTimer) {
	QEventLoop_ProcessEvents3(this.h, flags, deadline.cPointer())
}

func (this *QEventLoop) Exec() int {
	return (int)(QEventLoop_Exec(this.h))
}

func (this *QEventLoop) IsRunning() bool {
	return (bool)(QEventLoop_IsRunning(this.h))
}

func (this *QEventLoop) WakeUp() {
	QEventLoop_WakeUp(this.h)
}

func (this *QEventLoop) Event(event *QEvent) bool {
	return (bool)(QEventLoop_Event(this.h, event.cPointer()))
}

func (this *QEventLoop) Exit() {
	QEventLoop_Exit(this.h)
}

func (this *QEventLoop) Quit() {
	QEventLoop_Quit(this.h)
}

func QEventLoop_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QEventLoop_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QEventLoop_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QEventLoop_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QEventLoop) ProcessEvents1(flags ProcessEventsFlags) bool {
	return (bool)(QEventLoop_ProcessEvents1(this.h, flags))
}

func (this *QEventLoop) Exec1(flags ProcessEventsFlags) int {
	return (int)(QEventLoop_Exec1(this.h, flags))
}

func (this *QEventLoop) Exit1(returnCode int) {
	QEventLoop_Exit1(this.h, (int)(returnCode))
}

func (this *QEventLoop) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QEventLoop_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QEventLoop) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QEventLoop_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QEventLoop_MetaObject
func miqt_exec_callback_QEventLoop_MetaObject(self QEventLoop, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QEventLoop{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QEventLoop) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QEventLoop_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QEventLoop) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QEventLoop_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QEventLoop_Metacast
func miqt_exec_callback_QEventLoop_Metacast(self QEventLoop, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}
	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)
	virtualReturn := gofunc((&QEventLoop{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

type QEventLoopLocker struct {
	h          uintptr
	isSubclass bool
}

// NewQEventLoopLocker constructs a new QEventLoopLocker object.
func NewQEventLoopLocker() *QEventLoopLocker {
	g := newQEventLoopLocker(QEventLoopLocker_new())
	g.isSubclass = true
	return g
}

// NewQEventLoopLocker2 constructs a new QEventLoopLocker object.
func NewQEventLoopLocker2(loop *QEventLoop) *QEventLoopLocker {
	g := newQEventLoopLocker(QEventLoopLocker_new2(loop.cPointer()))
	g.isSubclass = true
	return g
}

// NewQEventLoopLocker3 constructs a new QEventLoopLocker object.
func NewQEventLoopLocker3(thread *QThread) *QEventLoopLocker {
	g := newQEventLoopLocker(QEventLoopLocker_new3(thread.cPointer()))
	g.isSubclass = true
	return g
}

func (this *QEventLoopLocker) Swap(other *QEventLoopLocker) {
	QEventLoopLocker_Swap(this.h, other.cPointer())
}

File: qt/gen_qeventloop.cpp
// +build ignore

#include <QDeadlineTimer>
#include <QEvent>
#include <QEventLoop>
#include <QEventLoopLocker>
#include <QMetaObject>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QThread>
#include <qeventloop.h>
#include "gen_qeventloop.h"

class MiqtVirtualQEventLoop : public virtual QEventLoop {
public:

	MiqtVirtualQEventLoop(): QEventLoop() {};
	MiqtVirtualQEventLoop(QObject* parent): QEventLoop(parent) {};

	virtual ~MiqtVirtualQEventLoop() = default;

	// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;

	// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
		if (handle__MetaObject == 0) {
			return QEventLoop::metaObject();
		}
		

		QMetaObject* callback_return_value = miqt_exec_callback_QEventLoop_MetaObject(const_cast<MiqtVirtualQEventLoop*>(this), handle__MetaObject);

		return callback_return_value;
	}

	// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QEventLoop::metaObject();

	}

	// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;

	// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
		if (handle__Metacast == 0) {
			return QEventLoop::qt_metacast(param1);
		}
		
		const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QEventLoop_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}

	// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QEventLoop::qt_metacast(param1);

	}

};

QEventLoop* QEventLoop_new() {
	return new MiqtVirtualQEventLoop();
}

QEventLoop* QEventLoop_new2(QObject* parent) {
	return new MiqtVirtualQEventLoop(parent);
}

void QEventLoop_virtbase(QEventLoop* src, QObject** outptr_QObject) {
	*outptr_QObject = static_cast<QObject*>(src);
}

QMetaObject* QEventLoop_MetaObject(const QEventLoop* self) {
	return (QMetaObject*) self->metaObject();
}

void* QEventLoop_Metacast(QEventLoop* self, const char* param1) {
	return self->qt_metacast(param1);
}

struct miqt_string QEventLoop_Tr(const char* s) {
	QString _ret = QEventLoop::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

bool QEventLoop_ProcessEvents(QEventLoop* self) {
	return self->processEvents();
}

void QEventLoop_ProcessEvents2(QEventLoop* self, ProcessEventsFlags flags, int maximumTime) {
	self->processEvents(flags, static_cast<int>(maximumTime));
}

void QEventLoop_ProcessEvents3(QEventLoop* self, ProcessEventsFlags flags, QDeadlineTimer* deadline) {
	self->processEvents(flags, *deadline);
}

int QEventLoop_Exec(QEventLoop* self) {
	return self->exec();
}

bool QEventLoop_IsRunning(const QEventLoop* self) {
	return self->isRunning();
}

void QEventLoop_WakeUp(QEventLoop* self) {
	self->wakeUp();
}

bool QEventLoop_Event(QEventLoop* self, QEvent* event) {
	return self->event(event);
}

void QEventLoop_Exit(QEventLoop* self) {
	self->exit();
}

void QEventLoop_Quit(QEventLoop* self) {
	self->quit();
}

struct miqt_string QEventLoop_Tr2(const char* s, const char* c) {
	QString _ret = QEventLoop::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

struct miqt_string QEventLoop_Tr3(const char* s, const char* c, int n) {
	QString _ret = QEventLoop::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

bool QEventLoop_ProcessEvents1(QEventLoop* self, ProcessEventsFlags flags) {
	return self->processEvents(flags);
}

int QEventLoop_Exec1(QEventLoop* self, ProcessEventsFlags flags) {
	return self->exec(flags);
}

void QEventLoop_Exit1(QEventLoop* self, int returnCode) {
	self->exit(static_cast<int>(returnCode));
}

void QEventLoop_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQEventLoop*>( (QEventLoop*)(self) )->handle__MetaObject = slot;
}

QMetaObject* QEventLoop_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQEventLoop*)(self) )->virtualbase_MetaObject();
}

void QEventLoop_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQEventLoop*>( (QEventLoop*)(self) )->handle__Metacast = slot;
}

void* QEventLoop_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQEventLoop*)(self) )->virtualbase_Metacast(param1);
}

void QEventLoop_Delete(QEventLoop* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQEventLoop*>( self );
	} else {
		delete self;
	}
}

QEventLoopLocker* QEventLoopLocker_new() {
	return new QEventLoopLocker();
}

QEventLoopLocker* QEventLoopLocker_new2(QEventLoop* loop) {
	return new QEventLoopLocker(loop);
}

QEventLoopLocker* QEventLoopLocker_new3(QThread* thread) {
	return new QEventLoopLocker(thread);
}

void QEventLoopLocker_Swap(QEventLoopLocker* self, QEventLoopLocker* other) {
	self->swap(*other);
}

void QEventLoopLocker_Delete(QEventLoopLocker* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QEventLoopLocker*>( self );
	} else {
		delete self;
	}
}

